lib/main.js

```diff
@@ -10,139 +10,176 @@ import { CompositeDisposable } from 'atom';
+      this.messageFormat = value;
+    }));
+    this.subscriptions.add(atom.config.observe('linter-pylint.pythonPath', (value) => {
+      this.pythonPath = value;
+    }));
+    this.subscriptions.add(atom.config.observe('linter-pylint.workingDirectory', (value) => {
+      this.workingDirectory = value.replace(delimiter, '');
+    }));
+    this.subscriptions.add(atom.config.observe('linter-pylint.disableTimeout', (value) => {
+      this.disableTimeout = value;
+    }));
+  },
+
+  deactivate() {
+    this.subscriptions.dispose();
+  },
+
+  provideLinter() {
+    return {
+      name: 'Pylint',
+      grammarScopes: ['source.python', 'source.python.django'],
+      scope: 'file',
+      lintsOnChange: false,
+      lint: async (editor) => {
+        const filePath = editor.getPath();
+        const fileDir = dirname(filePath);
+        const fileText = editor.getText();
+        const projectDir = getProjectDir(filePath);
+        const cwd = fixPathString(this.workingDirectory, fileDir, projectDir);
+        const execPath = fixPathString(this.executablePath, '', projectDir);
+        let format = this.messageFormat;
+        const patterns = {
+          '%m': 'msg',
+          '%i': 'msg_id',
+          '%s': 'symbol',
+        };
+        Object.keys(patterns).forEach((pattern) => {
+          format = format.replace(new RegExp(pattern, 'g'), `{${patterns[pattern]}}`);
         });
-      });
-    },
-  };
-}
+        const env = Object.create(process.env, {
+          PYTHONPATH: {
+            value: [
+              process.env.PYTHONPATH, fileDir, projectDir,
+              fixPathString(this.pythonPath, fileDir, projectDir),
+            ].filter(x => !!x).join(delimiter),
+            enumerable: true,
+          },
+          LANG: { value: 'en_US.UTF-8', enumerable: true },
+        });
+
+        const args = [
+          `--msg-template='{line},{column},{category},{msg_id}:${format}'`,
+          '--reports=n',
+          '--output-format=text',
+        ];
+        if (this.rcFile !== '') {
+          args.push(`--rcfile=${fixPathString(this.rcFile, fileDir, projectDir)}`);
+        }
+        args.push(filePath);
+
+        const execOpts = { env, cwd, stream: 'both' };
+        if (this.disableTimeout) {
+          execOpts.timeout = Infinity;
+        }
+
+        const data = await exec(execPath, args, execOpts);
+
+        if (editor.getText() !== fileText) {
+          // Editor text was modified since the lint was triggered, tell Linter not to update
+          return null;
+        }
+
+        const filteredErrors = filterWhitelistedErrors(data.stderr);
+        if (filteredErrors) {
+          // pylint threw an error we aren't ignoring!
+          throw new Error(filteredErrors);
+        }
+
+        const lineRegex = /(\d+),(\d+),(\w+),(\w\d+):(.*)\r?(?:\n|$)/g;
+        const toReturn = [];
+
+        let match = lineRegex.exec(data.stdout);
```

Comment:
Duplication here and at the end of while loop, What about `while ((match = lineRegex.exec(data.stdout)) !== null)` while it looks a little ugly, it saves us from duplicates, pure personal taste tho so your call